File: HTML.cpp
//Dongyao Zhu

#include "HTML.hpp"

/*
 * name: constructor
 * description: builds the HTML tree structure from the tags
        row by row, each row determined by a reference tag that 
        starts most topleft and spans most row.
 * input: size of window, a priority queue of tags, given in the generator.cpp
 * output: a HTML tree
 */
HTML :: HTML(double width, double height, 
    priority_queue<Tag *, vector<Tag *>, TagPtrComp> & tags){

    root = new Tag(0, 0, width, height, "body", "", 
        "\tbackground-color: #293030;\n");
    double last = 0;
    //while all not finished
    while(!tags.empty()){
        Tag * current = tags.top();
        //this pq will store all tags in this row
        priority_queue<Tag *, vector<Tag *>, sortS> row;
        tags.pop();
        //while this row not finished
        while(!tags.empty()){
            Tag * next = tags.top();
            //next is below(it is in next row), moving to next row
            if(current -> botRightY() <= next -> topLeftY()){
                break;
            }else{
                //valid relations: next tag is NOT BELOW reference, which
                //may need to include more rows if next protrudes downwards
                current = current -> expandRow(next);
                row.push(next);
                tags.pop(); 
            }
        }//end of while
        row.push(current);
        //wrap row elements 2 by 2 then push back, think of huffman encoding
        //note: not really necessary, but will make html structure more clear
        while(row.size() > 1){
            Tag * a = row.top();
            row.pop();
            Tag * b = row.top();
            row.pop();
            row.push(Tag :: wrap(a, b));
        }
        //the final gigantic wrapped element
        if(row.size() == 1){
            Tag * t = row.top();
            //see Tag for relative position math
            double moveY = t -> topLeftY() - last;
            t -> style += "\ttop: " + to_string(moveY) + "%;\n";
            t -> style += "\tleft: " + to_string(t -> topLeftX()) + "%;\n";
            //total vertical adjustments
            last = t -> botRightY() - (moveY > 0 ? moveY : 0);
            t -> parent = root;
            root -> children.push_back(t);
            row.pop();
        }
    }//end of while
}//HTML

/*
 * name: destructor
 * description: delegate to Tag's destructor
 * input/output: none
 */
HTML :: ~HTML(){
    delete root;
}

/*
 * name: write
 * description: generate HTML and CSS source code from html structure tree, DFS
 * input: a designated path to store the file, a string css file path
 * output: whether successfully written a html and css file to the path
 */
bool HTML :: write(string pathHTML, string pathCSS){
    ofstream html(pathHTML);
    ofstream css(pathCSS);
    if(!root || !html.is_open() || !css.is_open()){
        return false;
    }
    writeHeader(html, pathCSS);
    writeSize(css);
    string tab = "\t";
    for(Tag * t : root -> children){
        helper(html, css, t, tab);
    }
    html << "</body>\n</html>\n";
    html.close();
    css.close();
    return true;
}

void HTML :: addStyle(string pathCSS){
    moreCSS += "\t<link rel=\"stylesheet\" type=\"text/css\" href=\"" + 
        pathCSS + "\">\n";
}

/*
 * name: writeHeader
 * description: writes out necessary HTML beginning information and css links
 * input: an ofstream & to write, a string of css link
 * output: none
 */
void HTML :: writeHeader(ofstream & html, string pathCSS){
    html << header << moreCSS << css << pathCSS << "\">\n</head>\n<body>\n";
}

/*
 * name: writeSize
 * description: writes webpage size to CSS
 * input: ofstream &
 * output: none
 */
void HTML :: writeSize(ofstream & css){
    css << "body{\n" << root -> style 
        << "\twidth: " << root -> W()
        << "vw;\n\theight: " << root -> H()
        << "vw;\n}\n";
}

/*
 * name: helper
 * description: write helper
 * input: same as write
 * output: none
 */
void HTML :: helper(ofstream & html, ofstream & css, 
    Tag * current, string tabs){

    html << tabs << current -> openTag();
    //write id(CHANGE TO ACTUAL CONTENT) except for wrappers
    if(current -> cls != "wrap"){
        html << tabs << "\t" << current -> id << endl;
    }
    //write style
    if(current -> style != ""){
        css << "#" << current -> id << "{\n" << current -> style
            << "\twidth: " << current -> wPct
            << "%;\n\theight: " << current -> hPct  
            << "%;\n}\n\n";
    }
    //running DFS for the html structure
    for(Tag * t : current -> children){
        helper(html, css, t, tabs + "\t");
    }
    html << tabs << current -> closeTag();
}

//eof

File: HTML.py

#Dongyao Zhu

from Tag import Tag
from heapq import heappush, heappop, heapify

class HTML():

	placeholder = ('Lorem ipsum dolor sit amet, pri nostrud '
				'scaevola at, ex agam habeo assueverit mei.')

	def __init__(self, raw):
		self.header = ('<!DOCTYPE html>\n<html>\n'
			'<head>\n\t<meta name = \"viewport\" content = \"'
			'width = device-width, initial-scale = 1, '
			'maximum-scale = 1.0, user-scalable = 0\"/>\n\t'
			'<link rel=\"stylesheet\" type=\"text/css\" href=\"style.css\"/>')
		self.root = Tag(0, 0, 100, 100, 'body', \
			'\tbackground-color: #293030;\n')
		self.last = 0
		raw = [Tag(i[0], i[1], i[2], i[3], i[4]) for i in raw]
		self.makeTree(raw)
		self.final = ''

	def makeTree(self, tags):
		'''
	    builds the HTML tree structure from the tags
	    row by row, each row determined by a reference tag that 
	    starts most topleft and spans most row.
	    '''
		heapify(tags)
		while len(tags) > 0:
			reference = heappop(tags)
			row = []
			#get row
			while len(tags) > 0:
				neighbour = tags[0];
				if reference.bry <= neighbour.tly:
					break
				else:
					reference = reference.expandRow(neighbour)
					heappush(row, neighbour)
					heappop(tags)
			heappush(row, reference)
			#merge row
			while len(row) > 1:
				a = heappop(row)
				b = heappop(row)
				heappush(row, Tag.wrap(a, b))
			if len(row) == 1:
				t = heappop(row)
				moveY = t.tly - self.last
				t.style += '\ttop: ' + str(moveY) + '%;\n'
				t.style += '\tleft: ' + str(t.tlx) + '%;\n'
				#total vertical adjustments
				self.last = t.bry - moveY if moveY > 0 else 0
				self.root.children.append(t)

	def toHTML(self):
		#generate HTML and CSS source code from html structure tree, DFS
		self.final += self.header + '\n<style>\n'
		self.final += 'body{\n' + self.root.style + \
			'\twidth: 100vw;\n\theight: 100vw;\n}\n'
		html, css, tab = [], [], '\t'
		#dfs, will crash if using python2
		for t in self.root.children:
			HTML.helper(html, css, t, tab);
		self.final += ''.join(css)
		self.final += '</style>\n</head>\n<body>\n'
		self.final += ''.join(html)
		self.final += ('</body>\n</html>\n')
		return self.final

	def helper(html, css, t, tabs):
		#write helper
		html.append(tabs + t.openTag())
		if t.name == 'p' or t.name[0] == 'h':
			html.append('\n' + tabs + '\t' + HTML.placeholder + '\n' + tabs)
		if t.style != '':
			css.append('#' + t.id + '{\n' + t.style + '\twidth: ' \
				+ str(t.wPct) + '%;\n\theight: ' + str(t.hPct) + '%;\n}\n')
		if len(t.children) != 0:
			html.append('\n')
			for sub in t.children:
				HTML.helper(html, css, sub, tabs + '\t')
			html.append(tabs + t.closeTag())
		else:
			html.append(t.closeTag())
